File: compiler_f/NQueens/src/nq-serial.f90
!#==============================================================================
!#
!#  SAMPLE SOURCE CODE - SUBJECT TO THE TERMS OF SAMPLE CODE LICENSE AGREEMENT,
!#  http://software.intel.com/en-us/articles/intel-sample-source-code-license-agreement/
!#
!#  Copyright 2012-2016 Intel Corporation
!#
!#  THIS FILE IS PROVIDED "AS IS" WITH NO WARRANTIES, EXPRESS OR IMPLIED,
!#  INCLUDING BUT NOT LIMITED TO ANY IMPLIED WARRANTY OF MERCHANTABILITY,
!#  FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT OF INTELLECTUAL
!#  PROPERTY RIGHTS.
!#
!#==============================================================================
!#
!#
!#******************************************************************************
!# Content:
!#      NQueens - Serial sample program for Intel Visual Fortran
!#******************************************************************************
!

program NQueens

! Solve the nqueens problem - base serial version
!
! Original C++ code by Ralf Ratering & Mario Deilmann
! Fortran version by Steve Lionel
!
! Read http://en.wikipedia.org/wiki/Nqueens for background
!
! To set command line argument in Visual Studio, right click on the project name and select Properties.
! Under Debugging, enter the argument (board size) in Command Arguments.

implicit none

integer :: nrOfSolutions = 0 
integer :: size = 0

character(2) :: cmdarg
integer :: stat
integer :: time_start, time_end, count_rate

! Get command line argument
if (command_argument_count() < 1) then
  print *, "Usage: nqueens-serial boardSize"
  stop
  end if
call get_command_argument (1, cmdarg, status=stat)
if (stat /= 0) then
  print *, "Invalid boardSize"
  stop
  end if
read (cmdarg,*,iostat=stat) size
if ((stat /= 0) .or. (size < 1)) then
  print *, "Error: boardSize must be between 1 and 99"
  stop
  end if

! Begin
print 101, "Starting serial recursive solver for size ", size
101 format (A,I0,A)
call system_clock (time_start)
call solve (size)
call system_clock (time_end, count_rate)
print 101, "Number of solutions: ", nrOfSolutions
print 101, "Calculations took ", (time_end-time_start) / (count_rate/1000), "ms."

contains

! Routine to print the board

subroutine print (queens)
  implicit none
  integer, intent(in) :: queens(:)
  integer :: row, col

  do row=1,size
    do col=1,size
      if (queens(row) == col) then
        write (*,'(A)',advance='no') "Q"
      else
        write (*,'(A)',advance='no') "-"
      end if
    end do
  write (*,'(A)')
  end do
write (*,*)
end subroutine print

! Recursive routine to set a queen on the board

recursive subroutine setQueen (queens, row, col)
  implicit none
  integer, intent(inout) :: queens(:)
  integer, intent(in) :: row, col
  integer :: i
  
  do i=1,row-1
    ! vertical attacks
    if (queens(i) == col) return
    ! diagonal attacks
    if (abs(queens(i)-col) == (row-i)) return
    end do
    
  ! column is ok, set the queen
  queens(row) = col
  
  if (row == size) then
    nrOfSolutions = nrOfSolutions + 1
!DEC$ IF DEFINED(PRINT)
    call print(queens)
!DEC$ END IF
  else
    ! try to fill next row
    do i=1,size
      call setQueen (queens, row+1, i)
    end do
  end if
end subroutine SetQueen

! Main solver routine

subroutine solve (size)
  implicit none
  integer, intent(in):: size
  integer, allocatable :: lcl_queens(:)
  integer i

  do i=1,size
    ! try all positions in first row
!DEC$ IF DEFINED(PRINT)
    print '(A,I0)', "Call SetQueen with: row:1, column: ", i
!DEC$ END IF
    allocate(lcl_queens(size),source=0)
    call SetQueen (lcl_queens, 1, i)
    deallocate(lcl_queens)
  end do
end subroutine solve

end program nQueens

File: compiler_c/nqueens_samples/src/common/nq-serial.cpp
//==============================================================
//
// SAMPLE SOURCE CODE - SUBJECT TO THE TERMS OF SAMPLE CODE LICENSE AGREEMENT,
// http://software.intel.com/en-us/articles/intel-sample-source-code-license-agreement/
//
// Copyright Intel Corporation
//
// THIS FILE IS PROVIDED "AS IS" WITH NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING BUT
// NOT LIMITED TO ANY IMPLIED WARRANTY OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
// PURPOSE, NON-INFRINGEMENT OF INTELLECTUAL PROPERTY RIGHTS.
//
// =============================================================
// [DESCRIPTION]
// Solve the NQueens problem, serial version
// 
// [RUN]
// From the command line:
//      nq-serial.exe boardSize
// where boardSize is the size of the board, in the range 4-20
//
// [EXPECTED OUTPUT]
// Board Size   Number of Solutions
//     4                2
//     5               10
//     6                4
//     7               40
//     8               92
//     9              352
//    10              724
//    11             2680
//    12            14200
//    13            73712
//    14           365596
//    15          2279184
//=============================================================

#include "stdafx.h"
#include <mmsystem.h>
#include <stdlib.h>
#include "nq-serial.h"

CString GetStampString()
{
	SYSTEMTIME curTime;
    GetLocalTime(&curTime);

	CString strOutText(""); 
	strOutText.AppendFormat(L"%02d/%02d, %02d:%02d:%02d", curTime.wMonth, curTime.wDay, curTime.wHour, curTime.wMinute, curTime.wSecond);

	return strOutText; 
}

void nqSerialSetQueen(int queens[], int row, int col) 
{
	int i = 0;
	for (i=0; i<row; i++) {
		// vertical attacks
		if (queens[i]==col)
			return;
		// diagonal attacks
		if (abs(queens[i]-col) == (row-i) )
			return;
	}

	// column is ok, set the queen
	queens[row]=col;

	if (row==g_nsize-1)
	{
		g_nSolutions++;
	}
	else {
		// try to fill next row
		for (i=0; i<g_nsize; i++)
		  nqSerialSetQueen(queens, row+1, i);
	}
}



void nqSerialSolve(int nsize)
{
	g_nsize = nsize; 


	for(int i=0; i<g_nsize; i++)
	{
		// create separate array for each recursion
		int* pNQ = new int[g_nsize];

		// try all positions in first row
		nqSerialSetQueen(pNQ, 0, i);

		delete pNQ;
	}

}


CString nqSerialGetSolutions(int nSize)
{
	CString strStamp = GetStampString();
	CString strOutText("Starting serial recursive solver for size "); 
	strOutText.AppendFormat(L"%d, starting %s\r\n", nSize, strStamp);

	DWORD startTime=timeGetTime();

	// reset g_nSolutions
	g_nSolutions = 0;

	// calc the solutions
	nqSerialSolve(nSize);

	DWORD dwTimeUsed = timeGetTime()-startTime;

	strOutText.AppendFormat(L"    Number of solutions: %d\r\n    Calculations took: %u ms.\r\n\r\n", g_nSolutions, dwTimeUsed);

	return strOutText;
}
